plugins/pam.lua

```diff
@@ -0,0 +1,264 @@
+--   User Attributes: uidnumber, gidnumber
+--   Group Attributes: gidnumber
+--   Groups: pam_users (default)
+--
+-- Strives to maintain:
+--   Users: all have a unique uidnumber
+--   Users: all users by default have a gidnumber corresponding to the
+--          pam_users group created.
+--   Groups: all groups have a unique uidnumber
+--
+-- Group will only be created when a user needs the gidnumber for
+-- it, this is done to avoid it being created as group 1, before
+-- lldap_admin, in new setups.
+--
+-- TODO:
+--   3. max uid should be stored in a key-value store instead
+--
+
+-- keys used for this plugins key-value store
+local KVKEYS = {
+    NEXT_UID = "next-uid"
+}
+
+local get_config_gid_offset = function(context)
+    if context.configuration.gid_offset ~= nil then
+        return tonumber(context.configuration.gid_offset)
+    else
+        -- return default
+        return 100000
+    end
+end
+
+
+local get_config_uid_offset = function(context)
+    if context.configuration.uid_offset ~= nil then
+        return tonumber(context.configuration.uid_offset)
+    else
+        -- return default
+        return 100000
+    end
+end
+
+local get_config_default_group = function(context)
+    if context.configuration.default_group ~= nil then
+        return context.configuration.default_group
+    else
+        -- return default
+        return "pam_users"
+    end
+end
+
+local resolve_gid_default_group = function(context)
+    local group_name = get_config_default_group(context)
+    local groups = context.api:list_groups({ filter = "(cn=" .. group_name .. ")" })
+    local group_id = nil
+    if lldap.tables:empty(groups) then
+        lldap.log:debug("Creating group:" .. group_name)
+        local id, err = context.api:create_group({
+            display_name = group_name,
+            attributes = {}
+        })
+        if err ~= nil then
+            -- Error
+            error("Failed to create group: " .. group_name, 1)
+        end
+        lldap.log:debug("Created '" .. group_name .. "' group with id " .. tostring(id))
+        local group_gidnumber = get_config_gid_offset(context) + id
+        local res, err = context.api:update_group({
+            group_id = id,
+            insert_attributes = {
+                gidnumber = { int = group_gidnumber }
+            }
+        })
+        if err ~= nil then
+            error("Failed to set gidnumber attribute to '" .. group_name .. "' group", 1)
+        end
+        lldap.log:debug("Assigned gidnumber to '" .. group_name .. "': " .. tostring(group_gidnumber))
+        group_id = group_gidnumber
+    else
+        group_id = groups[1].attributes.gidnumber.int
+    end
+    return group_id
+end
+
+local resolve_next_uid = function(context, uid_offset)
+    local nextuid, err = context.kvstore:fetch_and_increment(KVKEYS.NEXT_UID, uid_offset)
+    if err ~= nil then
+        error("Unable to resolve next-uid from plugin KV store. Error: " .. tostring(err), 1)
```

Comment:
I guess that's a question of lua style, but I think in that case the idiomatic thing would be to return 2 values from the function, go-style, instead of logging. Here you'd be returning `nil, "Unable..."` and below just `return nextuid` (the second `nil` is implicit when unpacking the result at the callsite)

Reply:
And actually, that error is not handled in the callsite, and will lead to some weirdness, no?

Reply:
> And actually, that error is not handled in the callsite, and will lead to some weirdness, no?

Yeah, the intention was actually to abort the script -- not meaning it had been fully thought through though. Writing this was what made me make the todo on error return values from plugin :-)